CLOUDP-279933: documentation (#462)
mongodb/atlas-sdk-go: docs/doc_4_authentication.md

# Authenticating with the Atlas Go SDK

The `atlas-sdk-go` library utilizes Digest authentication as its default authentication method.
You can [create an API key](https://www.mongodb.com/docs/atlas/configure-api-access/#create-an-api-key-in-an-organization) through the Atlas UI or the Atlas CLI.

> To learn more about API authentication, refer to the [Atlas Administration API Authentication](https://www.mongodb.com/docs/atlas/api/api-authentication).

Reviewer: ```suggestion
To learn more about API authentication, refer to [Atlas Administration API Authentication](https://www.mongodb.com/docs/atlas/api/api-authentication).
```

Looks like I missed this one earlier. Our convention when referring to pages is to just use their name without an article. One exception is if we are linking to an overarching page that serves as an entrypoint to a specific area of the documentation, and we are explicitly using this as a proxy for that whole area. For example, if we were to link to the [Atlas Search landing page](https://www.mongodb.com/docs/atlas/atlas-search/) intending to suggest that the user read Atlas Search docs generally, we would say "refer to **the** Atlas Search documentation".

	fmt.Printf("Projects: %v\n", projects)
}
```

## (Preview) Using the Atlas Go SDK with Service Account Authentication

Atlas SDK Go provides OAuth Authentication using Service Accounts (currently available as a [Preview](https://www.mongodb.com/resources/beta-programs) feature)

> Service Account implement OAuth [client_credentials](https://oauth.net/2/grant-types/client-credentials) grant.

Reviewer: "Service Account" is, I gather, not a collective term. Our possible alternatives are: 

```suggestion
Service Accounts implement OAuth [client_credentials](https://oauth.net/2/grant-types/client-credentials) grants.
```

OR

```suggestion
A Service Account implements an OAuth [client_credentials](https://oauth.net/2/grant-types/client-credentials) grant.
```

The latter approach makes more sense if you are going to continue with a description of features of a single Service Account, especially if that description is going to involve assumptions about one instance of a Service Account (e.g. specific values for parameters). In this case, we are aiming at a higher-level conceptual summary, so I recommend the former approach that uses plurals.

> For more information about feature please refer to [Service Account Public documentation.](https://www.mongodb.com/docs/atlas/api/service-accounts-overview/)

Author: This PR should not be merged before link is active.

## OAuth Authentication

### Authenticating with OAuth ClientID and ClientSecret

### Admin API Authentication using Service Accounts

Authenticating using Service Accounts

```go
package main

import (
	"log"
	"os"

	"go.mongodb.org/atlas-sdk/v20241023001/admin"
)

func main() {
	clientID := os.Getenv("MONGODB_ATLAS_CLIENT_ID")
	clientSecret := os.Getenv("MONGODB_ATLAS_CLIENT_SECRET")

	if clientID == "" || clientSecret == "" {
		log.Fatal("Missing CLIENT_ID or CLIENT_SECRET environment variables")
	}

	// Using ClientID and ClientSecret. No cache supported (nil).
	sdk, err := admin.NewClient(admin.UseOAuthAuth(clientID, clientSecret, nil))

	// Make API calls
}
```

> NOTE: This method of initialization provides **No Ability to Revoke API**: once a token is issued, it cannot be invalidated until it expires.

### Specifying a Token Cache

In this example, we will demonstrate how to use the OAuth Client Credentials flow with a custom token cache.

> The cache allows you to store OAuth tokens for reuse across application restarts, improving efficiency by reducing the number of token requests to the authorization server.
> 
> **Note:** Cache prevents from requesting multiple tokens which are subject to OAuth Token Limits and Rate Limiting.

Reviewer: ```suggestion
The cache allows you to store OAuth tokens for reuse across application restarts, improving efficiency by reducing the number of token requests to the authorization server. This reduction in requests also minimizes the impact of OAuth Token Limits and Rate Limiting.
```

I think the word 'minimizing' is important here; we probably can't provide a 100% guarantee that the user won't ever run afoul of these limits, so we should make it clear that while this sharply mitigates the concern, it doesn't eliminate it. Also, folding that statement into the existing paragraph spares us some redundancy in the note.

	"go.mongodb.org/atlas-sdk/v20241023001/admin"
)

type MyTokenCache struct {
	fileContent []byte
	mu          sync.Mutex
}

func (s *MyTokenCache) RetrieveToken(_ context.Context) (string, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	var tkn string
	err := json.Unmarshal(s.fileContent, &tkn)
	if err != nil {
		return "", err
	}

	return tkn, nil
}

func (s *MyTokenCache) SaveToken(_ context.Context, tkn string) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	fileData, err := json.Marshal(tkn)
	if err != nil {
		return err
	}
	s.fileContent = fileData
	return nil
}

func main() {
	clientID := os.Getenv("MONGODB_ATLAS_CLIENT_ID")
	clientSecret := os.Getenv("MONGODB_ATLAS_CLIENT_SECRET")

	if clientID == "" || clientSecret == "" {
		log.Fatal("Missing CLIENT_ID or CLIENT_SECRET environment variables")
	}

	fileTokenCache := MyTokenCache{}
	// Using ClientID and ClientSecret. No cache supported (nil).
	httpClient, err := admin.NewClient(admin.UseOAuthAuth(clientID, clientSecret, fileTokenCache))
	sdk, err := admin.NewClient(admin.UseHTTPClient(httpClient))

	// Make requests to the API
	// ...
	// Revoke Token
	_ = tokenSource.RevokeToken()
}
```

## Advanced Use Cases

### Revocation

> Revocation invalidates the token before its expiration date.
> Revocation should be used to "logg out" the current OAuth client and configuring a new one.

Reviewer: ```suggestion
Revocation invalidates an OAuth token before its expiration date. This effectively "logs out" the current OAuth client and allows you to configure a new client.
```

Generally, at the start of a new section, it's best to be very explicit (i.e. "an OAuth token" vs. "the token"). An exception to this case is when you are writing a strict sequence of events and "the token" can be unambiguously taken to refer to a specific token created previously in that sequence.

Also, rather than using the suggestive "should", let's stick to telling the user what something does or what is possible, unless we are making explicit recommendations from among alternatives.

```go
// Sounding code omitted for brevity 
err := tokenSource.RevokeToken()
if err != nil {
	log.Fatalf("Error: %v", err)
}
```

### Creating a Custom Transport

You can create a custom transport to inject the OAuth token into HTTP requests. 

> The `OAuthCustomHTTPTransport` provides an `UnderlyingTransport` field that specifies the transport to be used for requests.

Reviewer: ```suggestion
The `OAuthCustomHTTPTransport` provides an `UnderlyingTransport` field that specifies the transport to use for requests.
```

Prefer the active voice where reasonable.

```go
tokenSource := credentials.NewTokenSourceWithOptions(credentials.AtlasTokenSourceOptions{
	ClientID:     clientID,
	ClientSecret: clientSecret,
})

transport := OAuthCustomHTTPTransport{
	UnderlyingTransport: <your_transport>,
	TokenSource:        tokenSource,
}
// Use transport with your own HTTP client or create a new one.
```

### Overriding the User Agent

> You can override the default user agent by specifying it in the `AtlasTokenSourceOptions`.

Reviewer: ```suggestion
You can override the default user agent by specifying an agent of your choice in the `AtlasTokenSourceOptions`.
```
Let's be explicit about what needs to be specified to achieve the override.
